source/common/buffer/watermark_buffer.cc

```diff
@@ -70,9 +70,16 @@ Api::IoCallUint64Result WatermarkBuffer::write(Network::IoHandle& io_handle) {
 
 void WatermarkBuffer::setWatermarks(uint32_t low_watermark, uint32_t high_watermark) {
   ASSERT(low_watermark < high_watermark || (high_watermark == 0 && low_watermark == 0));
+  if (overflow_watermark_multiplier_ > 0 &&
+      (overflow_watermark_multiplier_ * high_watermark) < high_watermark) {
```

Comment:
this comparison may result in false negatives.  It may be worth comparing high_watermark to numeric_limits<uint32_t>::max() / overflow_watermark_multiplier_ instead.